app/uk/gov/hmrc/fileupload/s3/S3JavaSdkService.scala

```diff
@@ -229,7 +236,82 @@ class S3JavaSdkService(configuration: com.typesafe.config.Config, metrics: Metri
       val outcome = Try(s3Client.deleteVersion(bucketName, summary.getKey, summary.getVersionId))
       Logger.info(s"Outcome of deleting $key / ${summary.getVersionId}: $outcome")
     }
+  }
+
+  override def zipAndPresign(
+    envelopeId: EnvelopeId,
+    files     : List[(FileId, Option[String])]
+  )(implicit
+    ec          : ExecutionContext,
+    materializer: Materializer
+  ): Future[ZipData] = {
+    val fileName = s"$envelopeId.zip"
+    val tempFile = TemporaryFile(prefix = "zip")
+    (for {
+       _            <- zipToFile(envelopeId, files, tempFile.file)
+       fileSize     =  tempFile.file.length
+       is           =  new java.io.FileInputStream(tempFile.file)
+
+       // decorate inputstream so we can calculate checksum on same pass
+       md           =  MessageDigest.getInstance("MD5")
+       dis          =  new DigestInputStream(is, md)
+
+       uploadResult <- uploadFile(
+                         bucketName = awsConfig.transientBucketName,
+                         key        = S3Key.forZipSubdir(awsConfig.zipSubdir)(fileName),
+                         file       = dis,
+                         metadata   = { val om = new ObjectMetadata()
+                                        om.setSSEAlgorithm(SSEAlgorithm.KMS.getAlgorithm)
+                                        om.setContentLength(fileSize)
+                                        om.setContentType("application/zip")
+                                        om
+                                      }
+                       )
+       url          =  presign(
+                         bucketName         = uploadResult.getBucketName,
+                         key                = uploadResult.getKey,
+                         expirationDuration = awsConfig.zipDuration
+                       )
+     } yield
+         ZipData(
+           name        = fileName,
+           size        = fileSize,
+           md5Checksum = Base64.getEncoder().encodeToString(md.digest()),
+           url         = url
+         )
+    ).andThen { case _ => tempFile.clean() }
+  }
 
+  def zipToFile(
+    envelopeId: EnvelopeId,
+    files     : List[(FileId, Option[String])],
+    targetFile: File
+  )(implicit materializer: Materializer): Future[akka.stream.IOResult] =
+    Source(
+      files.map { case (fileId, name) =>
+        val filename = name.getOrElse(UUID.randomUUID().toString)
+        download(
+          bucketName = awsConfig.transientBucketName,
+          key        = S3KeyName(S3Key.forEnvSubdir(awsConfig.envSubdir)(envelopeId, fileId))
+        ) match {
+          case None => sys.error(s"Could not find file $fileId, for envelope $envelopeId")
+          case Some(streamWithMetadata) => (ArchiveMetadata(filename), streamWithMetadata.stream)
+        }
+      }
+    ).via(Archive.zip())
```

Comment:
Just a small one, but were the outputted files identical between using Archive.zip() compared to the custom ZipStreamEnumerator?